packages/asset-swapper/src/network/samplers/dodo_v2.ts

```diff
@@ -0,0 +1,134 @@
+);
+
+const MAX_DODOV2_POOLS_QUERIED = 3;
+const DODO_V2_OFFSETS = [...new Array(MAX_DODOV2_POOLS_QUERIED)].map((_v, i) => new BigNumber(i))
+
+type SellContract = ERC20BridgeSamplerContract;
+type BuyContract = ERC20BridgeSamplerContract;
+type SellContractSellFunction = SellContract['sampleSellsFromDODOV2'];
+type BuyContractBuyFunction = BuyContract['sampleBuysFromDODOV2'];
+type SellContractSellFunctionParameters = Parameters<SellContractSellFunction>;
+type BuyContractBuyFunctionParameters = Parameters<BuyContractBuyFunction>;
+type SellContractSellFunctionReturn = UnwrapContractFunctionReturnType<ReturnType<SellContractSellFunction>>;
+type BuyContractBuyFunctionReturn = UnwrapContractFunctionReturnType<ReturnType<BuyContractBuyFunction>>;
+type SamplerSellEthCall = SamplerEthCall<DodoV2FillData,SellContractSellFunctionParameters,SellContractSellFunctionReturn>;
+type SamplerBuyEthCall = SamplerEthCall<DodoV2FillData,BuyContractBuyFunctionParameters,BuyContractBuyFunctionReturn>;
+
+export class DodoV2Sampler extends
+    OnChainSourceSampler<
+        SellContract,
+        BuyContract,
+        SellContractSellFunction,
+        BuyContractBuyFunction,
+        DodoV2FillData
+    >
+{
+    public static async createAsync(chain: Chain): Promise<DodoV2Sampler> {
+        return new DodoV2Sampler(
+            chain,
+            DODOV2_FACTORIES_BY_CHAIN_ID[chain.chainId],
+        );
+    }
+
+    protected constructor(chain: Chain, private readonly _factories: Address[]) {
+        super({
+            chain,
+            sellSamplerContractArtifactName: 'ERC20BridgeSampler',
+            buySamplerContractArtifactName: 'ERC20BridgeSampler',
+            sellSamplerContractType: ERC20BridgeSamplerContract,
+            buySamplerContractType: ERC20BridgeSamplerContract,
+            sellContractSellFunctionName: 'sampleSellsFromDODOV2',
+            buyContractBuyFunctionName: 'sampleBuysFromDODOV2',
+        });
+    }
+
+    public canConvertTokens(tokenAddressPath: Address[]): boolean {
+        return tokenAddressPath.length === 2;
+    }
+
+    protected async _getSellQuoteCallsAsync(
+        tokenAddressPath: string[],
+        takerFillAmounts: BigNumber[],
+    ): Promise<SamplerSellEthCall[]> {
+        const [takerToken, makerToken] = tokenAddressPath;
+        return this._factories.map(factory =>
+            DODO_V2_OFFSETS.map(offset => ({
+                args: [
+                    factory,
+                    offset,
+                    takerToken,
+                    makerToken,
+                    takerFillAmounts,
+                ],
+                getDexSamplesFromResult: ([isSellBase, poolAddress, samples]) => {
+                    return takerFillAmounts.map((a, i) => ({
+                        source: ERC20BridgeSource.DodoV2,
+                        fillData: { poolAddress, isSellBase },
+                        input: a,
+                        output: samples[i],
+                    }));
```

Comment:
Maybe we can have some helper magic defaults, which can be overridden in special cases 
```suggestion
                    return this.buildSamples(takerFillAmounts, samples, { poolAddress, isSellBase });
```

Does anything even really change over the sources here, maybe it's always just `fillData` construction ?

If we modify SwapRevertSampler to return the required encoded `fillData` then there might be even more cleanup available to us here.

Reply:
Idk I was also thinking of just fetching reserve amounts for uni2 and computing here but I guess that breaks with revert sampling

Reply:
Some (uni3) have different fill data per sample